src/lib/gadgets/gadgets.ts

```diff
@@ -474,6 +474,45 @@ const Gadgets = {
     div(x: Field3, y: Field3, f: bigint) {
       return ForeignField.div(x, y, f);
     },
+
+    /**
+     * Optimized multiplication of sums in a foreign field, for example: `(x - y)*z = a + b + c mod f`
+     *
+     * Note: This is much more efficient than using {@link ForeignField.add} and {@link ForeignField.sub} separately to
+     * compute the multiplication inputs and outputs, and then using {@link ForeignField.mul} to constrain the result.
+     *
+     * The sums passed into this gadgets are "lazy sums" created with {@link ForeignField.Sum}.
+     * You can also pass in plain {@link Field3} elements.
+     *
+     * **Assumptions**: The assumptions on the _summands_ are analogous to the assumptions described in {@link ForeignField.mul}:
+     * - each summand's limbs are in the range [0, 2^88)
+     * - summands that are part of a multiplication input satisfy `x[2] <= f[2]`
+     *
+     * @throws if the modulus is so large that the second assumption no longer suffices for validity of the multiplication.
+     * For small sums and moduli < 2^256, this will not fail.
+     *
+     * @throws if the provided multiplication result is not correct modulo f.
+     *
+     * @example
+     * ```ts
+     * // we assume that x, y, z, a, b, c are range-checked, analogous to `ForeignField.mul()`
+     * let xMinusY = ForeignField.Sum(x).sub(y);
+     * let aPlusBPlusC = ForeignField.Sum(a).add(b).add(c);
+     *
+     * // assert that (x - y)*z = a + b + c mod f
+     * ForeignField.assertMul(xMinusY, z, aPlusBPlusC, f);
```

Comment:
Maybe it would be helpful if you show what are the concrete gates that are created when calling `assertMul`. Well, maybe this is commented somewhere else, if so, I will find it while reviewing the other files of the PR.

Reply:
Ok, I posted an example constraint system here now: https://github.com/o1-labs/o1js/pull/1262#issuecomment-1823345067